sdk/storage/azdatalake/sas/service.go

```diff
@@ -0,0 +1,407 @@
+//go:build go1.18
+// +build go1.18
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package sas
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azdatalake/internal/exported"
+)
+
+// DatalakeSignatureValues is used to generate a Shared Access Signature (SAS) for an Azure Storage filesystem or path.
+// For more information on creating service sas, see https://docs.microsoft.com/rest/api/storageservices/constructing-a-service-sas
+// For more information on creating user delegation sas, see https://docs.microsoft.com/rest/api/storageservices/create-user-delegation-sas
+type DatalakeSignatureValues struct {
+	Version              string    `param:"sv"`  // If not specified, this defaults to Version
+	Protocol             Protocol  `param:"spr"` // See the Protocol* constants
+	StartTime            time.Time `param:"st"`  // Not specified if IsZero
+	ExpiryTime           time.Time `param:"se"`  // Not specified if IsZero
+	Permissions          string    `param:"sp"`  // Create by initializing FilesystemPermissions, FilePermissions or DirectoryPermissions and then call String()
+	IPRange              IPRange   `param:"sip"`
+	Identifier           string    `param:"si"`
+	FilesystemName       string
+	FilePath             string // Use "" to create a Filesystem SAS
+	DirectoryPath        string // Not nil for a directory SAS (ie sr=d)
+	CacheControl         string // rscc
+	ContentDisposition   string // rscd
+	ContentEncoding      string // rsce
+	ContentLanguage      string // rscl
+	ContentType          string // rsct
+	AuthorizedObjectID   string // saoid
+	UnauthorizedObjectID string // suoid
+	CorrelationID        string // scid
+}
+
+//TODO: add snapshot and versioning support in the future
+
+func getDirectoryDepth(path string) string {
+	if path == "" {
+		return ""
+	}
+	return fmt.Sprint(strings.Count(path, "/") + 1)
+}
+
+// SignWithSharedKey uses an account's SharedKeyCredential to sign this signature values to produce the proper SAS query parameters.
+func (v DatalakeSignatureValues) SignWithSharedKey(sharedKeyCredential *SharedKeyCredential) (QueryParameters, error) {
+	if v.ExpiryTime.IsZero() || v.Permissions == "" {
+		return QueryParameters{}, errors.New("service SAS is missing at least one of these: ExpiryTime or Permissions")
+	}
+
+	//Make sure the permission characters are in the correct order
+	perms, err := parsePathPermissions(v.Permissions)
```

Comment:
are we just parsing as path permissions since it is a super set of file system permissions?

Reply:
I call the method pathPermissions because both Dir and File permissions are identical, so this internal method is just used for both

Reply:
this also parses filesystem permissions right?

Reply:
I split the parser for paths and fs for future proofing, but yea the two essentially do the same thing

Reply:
I split the parsers of paths and fs just for future proofing, but yea they essentially do the same thing